### PR TITLE
Fix SingleChoiceInput accessibility issues

### PR DESCRIPTION
**Pull request by foxbunny**
- Replace SUI's Dropdown component with a custom Combobox component
- Implement a custom `<ind-combobox>` element used by Combobox
- Add `useNativeEvent` hook
- Tweak regform styles to handle the new controls

### CODE REVIEW

**Comment by micsucmed** [comment]
Hi @foxbunny, could you please explain in more detail what accessibility issues you're trying to solve here? I think you meant usability. There is still an issue when the user types in an entry that is not part of the drop-down options; there is no feedback that this entry is not valid and will not be sent (Given that the field is not required). And even if the field is required, the feedback gotten is "This field is required", which can be confusing since the user might think the field input has been provided, but in reality the input is not valid.
image

**Comment by foxbunny** [comment]
> please explain in more detail what accessibility issues you're trying to solve here

The primary accessibility issues that are being solved is incompatibility of the SUI combobox with screen readers. In particular, not announcing the options, not being clear about what effect the typing has, etc.

> There is still an issue when the user types in an entry that is not part of the drop-down options

I've already pointed this out to @ThiefMaster. It's a validation issue, not really an accessibility issue.

**Comment by tomasr8** [comment]
Thanks for the PR!

> I've already pointed this out to @ThiefMaster. It's a validation issue, not really an accessibility issue.

Can we do something about it in this PR?


Here's just some small styling issues I noticed:

When expanded I'd disable the bottom border radius on the input field:
image

I'd align the options with text input as it was before (i.e. increase left padding for the options)
image

The list needs a higher z-index:
image

From a usability point of view, I noticed that if I have two options, one uppercase and one lowercase e.g. `A` and `a`, when I type lowercase `a`, the `A` option is auto selected. That's not what I'd expect, especially since it changes the character I typed (from `a` to `A`)

**Comment by foxbunny** [comment]
> Can we do something about it in this PR?

My understanding was that it wasn't something we can do, but I can check the code, sure.

> Here's just some small styling issues I noticed:

Gotcha. Thanks.

> From a usability point of view, I noticed that if I have two options, one uppercase and one lowercase e.g. A and a

Hmmm. I'm not sure we should support this case at all. See, for screenreaders, there's absolutely no difference between capital A and lowercase a. So even if I make the combobox do case-sensitive matching in case of conflicts, it won't help screenreader users one bit.

**Comment by foxbunny** [comment]
> I can imagine regular users not being happy with the combobox changing their input.. If it doesn't hurt screenreader users I'd prefer case-sensitive matching

Do you really intend to have options that are identical except for the case? If you ask me, I'd forbid that during form creation.

**Comment by foxbunny** [comment]
Anyway, back to the discussion on case-matching. Suppose we have the following items in the list:

- ACME
- Acme
- acme
- aCme
- aCMe

The most trivial thing we can do is match at the start of the string case-sensitively. For example, if we type "A", it matches "ACME". If we type "Ac", it matches "Acme" and not "ACME". 

This has some edge cases. For example, if we type "acM" it will match "ACME" and not "acme" or "aCme" even though the first character is lower-case.

How elaborate do you want this algorithm to be?

**Comment by tomasr8** [comment]
> How elaborate do you want this algorithm to be?

A more radical idea, but what if we just don't autocomplete? As in, we keep the behaviour of the current dropdown?
My issue is purely with the fact that the input is actively changing what I'm typing, but maybe others have a different opinion? @ThiefMaster @micsucmed 

On an unrelated note:
- In the current dropdown one can search for any part of the text e.g. if you type `cme` you'll see `acme` as an option. I would prefer not to lose this functionality as it makes the search more useful
- A quick way to unselect the currently-selected option would be nice. The SUI dropdown has this `x` icon on the right side, maybe we could add it as well?

**Comment by foxbunny** [comment]
> In the current dropdown one can search for any part of the text e.g. if you type cme you'll see acme as an option. I would prefer not to lose this functionality as it makes the search more useful

I deliberately avoided using that approach because it causes usability issues such as matching something that isn't obviously what the user started off to type. (For instance in the country selector you start to type "Ser" it will give you "Montserrat" as the first option.) But anyway, I leave that up to you.

> A quick way to unselect the currently-selected option would be nice. The SUI dropdown has this x icon on the right side, maybe we could add it as well?

Sure, we can add that.

**Comment by foxbunny** [comment]
Updated the PR.

Fix styling issues, and implemented two strategies for handling typing based on the `aria-autocomplete` attribute, so now you can do both autocomplete and search (documented). Also added the clear button.

**Comment by foxbunny** [comment]
Rebased on master, and also fixed the validation logic so it correctly traps invalid choices.

**Comment by foxbunny** [comment]
@tomasr8 Reminder, just in case.

**Comment by tomasr8** [comment]
I noticed some differences from the old dropdown. Some of these might be nice to keep for the new one as well:

 - no results message:

obrazek
 
- I'd keep the same `x` icon:

obrazek
vs
obrazek

- When you have extra slots the dropdown to select the amount is much wider now (and the combobox changes width):

obrazek
vs
obrazek

- The hover bg color is the same as the one for the price now so it blends together:

obrazek
vs
obrazek

I'd also increase the height of the listbox a bit, currently it feels a bit too small

**Comment by foxbunny** [comment]
> When you have extra slots the dropdown to select the amount is much wider now (and the combobox changes width)

Do we even need the amount drop-down to be a combobox? I can replace it with a plain select list.

**Comment by foxbunny** [comment]
> no results message:

It would take a lot of additional code to get this feature in. I think there isn't such a dramatic difference between saying there are no matches and showing no matches from the UX standpoint, and therefore the additional work, maintenance burden, and complexity aren't justified. Of course, if you insist that this feature is crucial, I can still add it.

**Comment by tomasr8** [comment]
> > no results message:
> 
> It would take a lot of additional code to get this feature in. I think there isn't such a dramatic difference between saying there are no matches and showing no matches from the UX standpoint, and therefore the additional work, maintenance burden, and complexity aren't justified. Of course, if you insist that this feature is crucial, I can still add it.

Ok if it'd require a lot of extra work we can leave it out

> Do we even need the amount drop-down to be a combobox? I can replace it with a plain select list.

I think so :)

**Comment by foxbunny** [comment]
> I think so :)

Ok. I've pushed a fix for the combobox widths. Right now, these controls use different implementations (number of extra slots uses SUI dropdown). My plan is to replace the other ones once this PR is merged so we don't do too much in a single PR. Meanwhile you can adapt this to the time picker, for instance. Let me know if that works for you.

**Comment by tomasr8** [comment]
> Ok. I've pushed a fix for the combobox widths. Right now, these controls use different implementations (number of extra slots uses SUI dropdown). My plan is to replace the other ones once this PR is merged so we don't do too much in a single PR. Meanwhile you can adapt this to the time picker, for instance. Let me know if that works for you.

As long as the look is consistent across the different dropdowns it should be fine

File: indico/web/client/js/react/components/Combobox.jsx
Comment (ThiefMaster) [inline indico/web/client/js/react/components/Combobox.jsx]:
```suggestion
      <button type="button" value="clear" disabled={inputProps.disabled}>
```

Comment (ThiefMaster) [reply indico/web/client/js/react/components/Combobox.jsx]: I already applied this since it was a very simple fix and it worked fine when testing it locally.

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx
Comment (ThiefMaster) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:118]: I already added this since it worked fine when testing it locally.

File: indico/web/client/js/react/components/Combobox.jsx
Comment (ThiefMaster) [inline indico/web/client/js/react/components/Combobox.jsx:24]: This needs `autoComplete="off"` to avoid ugly cases such as this one:

image

**Comment by ThiefMaster** [comment]
@foxbunny I rebased this PR to the latest master and applied a few minor fixes for very straightforward things I encountered while testing.

I think finishing this PR could actually get higher priority on your side than the "datepicker for accommodation field" suggestion I made earlier today, since this one feels almost ready for a merge into master, and that way we'll have one less pending PR plus the advantage of being able to use the new Combobox (or should we call it `ComboBox`? ;)) elsewhere!

File: indico/web/client/js/react/components/Combobox.jsx
Comment (ThiefMaster) [inline indico/web/client/js/react/components/Combobox.jsx:42]: Is clicking this button supposed to open the dropdown? It feels strange because when I clear the field it typically means I want to to be empty, because in case I want to change the value I can simply open it again instead of clearing it first.

Comment (ThiefMaster) [inline indico/web/client/js/react/components/Combobox.jsx:42]: This button should not be displayed, if the field is `required`.

File: indico/web/client/js/custom_elements/ind_combobox.js
Comment (foxbunny) [reply indico/web/client/js/custom_elements/ind_combobox.js:1]: If it cannot be selected, maybe it makes more sense to simply remove it from the list? Or is the information that it's not longer available that important?

File: indico/web/client/js/react/components/Combobox.jsx
Comment (foxbunny) [reply indico/web/client/js/react/components/Combobox.jsx:42]: While fixing this, I noticed an interesting feature we got for free. If we set the input to `readonly`, then we get something that works like a select list. 😊

File: indico/web/client/js/custom_elements/ind_combobox.js
Comment (ThiefMaster) [reply indico/web/client/js/custom_elements/ind_combobox.js:1]: I think it's important to keep showing it. When buying tickets for something they usually also show sold-out options and don't simply make them disappear.

Comment (foxbunny) [reply indico/web/client/js/custom_elements/ind_combobox.js:1]: Gotcha.

**Comment by foxbunny** [comment]
@ThiefMaster I think I've taken care of the issues discussed here. I've also a minor accessibility issue with the price tag label not being associated with the extra slots field, so fixed that as well. Let me know if you spot anything else.

**Comment by tomasr8** [comment]
Thanks! I have some more style nitpicks ;)

- For disabled choices, I'd use the same lighter font color for the price as well (as it was previously).
Previous:
image
Current:
image

- The right alignment of the price label is a bit weird now (compare the first and second choice)
image

- I think the default height could be a bit larger. Currently it fits only 4 choices, which makes it feel a bit cramped

- I might be wrong but I don't think the clearable `X` button should be visible when the field is required:
image

**Comment by foxbunny** [comment]
@tomasr8 Fixed.

**Comment by tomasr8** [comment]
Thanks again!

> I think the default height could be a bit larger. Currently it fits only 4 choices, which makes it feel a bit cramped

Much better now! Do you think we could increase it up to 8 to match the old dropdown?

Would it be possible to not have the line wrapping on the price label?
Current:
image
Previous:
image

**Comment by foxbunny** [comment]
> Much better now! Do you think we could increase it up to 8 to match the old dropdown?

6~7 items is considered maximum for most humans for optimal operation.

> Would it be possible to not have the line wrapping on the price label?

Sure

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx
Comment (ThiefMaster) [inline indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:82]: When changing between two options there's a bug when 'extra slots' are enabled:

- The combobox value is not cleared, so if I switch from a max 3 slots choice (w/ 3 selected) to one that allows max 2 slots, the combobox still shows "3"
- Regardless of the choice's validity, when changing the option the combobox keeps the displayed value, which is now different from the submitted value. So If I change the choice like I mentioned in my previous point, it looks like I'm submitting the choice with 3 slots, but in reality I submit it with 1 slot.

Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:82]: Will check

Comment (ThiefMaster) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:82]: Likely related: When you use the `x` icon to clear the selected choice, the slots dropdown does not disappear.

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx
Comment (ThiefMaster) [inline indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:127]: Setting a value of `{"": -1}` seems wrong. It should be `{}` when nothing is selected. This is especially important if the field is not required, because in that case the "empty" value can actually get submitted.

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx
Comment (ThiefMaster) [inline indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: I don't like this lookup based on the caption... it's basically a roundtrip from unique id to (not guaranteed-to-be-unique) caption to unique id.

Can't we use the option value as the value of the ind-combobox? This would also match the standard HTML `<select>` `<option>` behavior more closely.

And this is functionality we'll certainly need to use this combobox elsewhere since there we typically don't want to have lots of boilerplate code around it to implement option values...

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx
Comment (ThiefMaster) [inline indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:126]: There's a bug when the field is not required and you type some nonsense in it: In that case there's no validation error, because it'll make no difference here if you explicitly cleared the field or wrote nonsense in it.

I think we need to add something similar here like what I did in the `FinalDatePicker`...

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx
Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: > I don't like this lookup based on the caption... it's basically a roundtrip from unique id to (not guaranteed-to-be-unique) caption to unique id.

That lookup has to happen somewhere anyway because this is *not* a select box. Then we again have this interpretation vs physical reality situation we had with "invalid" property.

The difference between combobox and a select box is that a combobox is fully editable, which means user can enter anything they want. If we don't want then, then we want to implement a different widget: searchable select list. That would mean a non-editable label coupled with a text field whose sole purpose is to filter the list.

As for uniqueness of the captions, if you ask me, that absolutely must be guaranteed in the management area, as having two identical captions with different values simply adds confusion for the user. How do you pick one if they are the same??

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx
Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:126]: Tbh, this is starting to sound like you don't want a combobox at all, but a searchable select list. Combobox really should be validated like any text field and then it should work as expected.

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx
Comment (ThiefMaster) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: > That lookup has to happen somewhere anyway because this is not a select box

Does it? In the end you create options from a list of options that typically have a label and a value. Couldn't the value of the selected option be stored in a member variable of the ind-combo-box instance or a hidden field? That way you'd never need to derive the value from the caption.

> As for uniqueness of the captions, if you ask me, that absolutely must be guaranteed in the management area, as having two identical captions with different values simply adds confusion for the user. How do you pick one if they are the same??
I agree that we typically do not want non-unique captions. But there are some cases - albeit very rare - where it could happen: Someone selects an option, then that option is removed by a manager (existing registrations keep it!), and later another option with the same caption is added back.

The dropdown will then show the old option (IIRC with an indicator next to the field that you may not be able to select it again if you deselect it and save), but also any new options since those would be valid choices now as well.

> The difference between combobox and a select box is that a combobox is fully editable, which means user can enter anything they want.

Maybe there was a misunderstanding, but that functionality never existed in Indico. At least not in the registration form (we use the SUI Dropdown with `allowAdditions` in some places, typically for an editable list of things like file extensions). The dropdown for the Single Choice registration form field was always supposed to be a (searchable) select list!

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx
Comment (ThiefMaster) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:126]: Yeah, this was never meant to be a combobox... (more details in my [other comment](https://github.com/indico/indico/pull/6310#discussion_r1646500737))

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx
Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: > Does it? In the end you create options from a list of options that typically have a label and a value.

The intended use case for combobox is basically exactly the same as using a `<datalist>`. It simply provides candidates for the text input. This is why it's allowed for the input to also perform the search. From the perspective of the end user, they are entering a value into the input. 

Select list is not editable because it does not allow arbitrary input. Since it is not editable, there is no expectation that an arbitrary value is allowed.

> Couldn't the value of the selected option be stored in a member variable of the ind-combo-box instance or a hidden field? That way you'd never need to derive the value from the caption.

That's technically doable (and that's similar to what SUI does), but it breaks the user's intuition about how the field works, and this is especially true for people who cannot see the field once they've left it. In order to correct the perception of what's to be expected of the field, the field itself cannot be editable. Therefore, a separate control is provided to facilitate searching and it's part of the popup listbox, not the form control.

> The dropdown will then show the old option (IIRC with an indicator next to the field that you may not be able to select it again if you deselect it and save), but also any new options since those would be valid choices now as well.

In that case, you can't use a combobox. It will be semantically very confused. I recommend using a select list with separate search in the listbox.

Comment (ThiefMaster) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: Then I think there was indeed some misunderstanding somewhere that made you go for a combobox :/

Anyway, I think we can still keep it around in case it becomes useful at some point elsewhere in Indico?

Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: Yeah, if you need an actual text input that supports completion candidates, then combobox will work as expected. I'll add a searchable select list to this PR and rewire the SingleSelectInput to use that instead of the combobox and keep the combobox custom element around.

Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: Before I do the PR, I'll implement a prototype in a sandbox, tho, and then we can take it from there.

Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: @ThiefMaster Check this out:

https://codepen.io/hayavuk/pen/mdYLYEL?editors=0110

It's still not 100% done yet, but it shows the behavior.

In general, it behaves more or less like the native select, with the differences being:

- It automatically expands the list when you press the arrow keys. I think that behavior is ok, but it's actually a hack to work around the issue of changes in value not being announced to screen readers when the list is closed.
- It has a search box.
- It can display arbitrary elements other than just the value label.

Comment (tomasr8) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: Works nice! I just got a question regarding the look of the element - are you planning to leave the search input inside the listbox? I'd like to keep the new select list similar to the combobox for uniformity and have the search input on top. What do you think?

Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: That doesn't make sense to me. How would you differentiate between selected value and random input?

Comment (tomasr8) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: We could just remove it when the field loses focus, like SUI does: https://react.semantic-ui.com/modules/dropdown/#types-search-selection

Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: That's the whole problem. You're not supposed to edit the field *after* it loses focus unless it's clear that the field is temporary (e.g., a filter for the list). This is where the whole thing about the original combobox started. SUI's combobox behavior is bad, I tried to fix it, and then it turned out it's bad because it's trying to present a searchable select list as a combobox.

Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: Also, I've updated the example to include a way to clear the field. You can add the `required` attribute to the `<x-select>` just like you would to any other control to see its behavior when the field is required.

Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: @tomasr8 I re-read your original comment, and I think this is basically what you had in mind:

https://codepen.io/hayavuk/pen/jOoomaN?editors=0100

This actually doesn't affect the separation between selected value caption and the search input. It simply visually places the input "on top" of the control caption so that it appears as if the caption is editable. I used the placeholder instead of populating the value as a work-around for the issues I mentioned above.

Comment (tomasr8) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: Just for reference, here's another example from MUI: https://mui.com/material-ui/react-autocomplete/#combo-box

Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:130]: Not sure what you mean by "add an extra value". Can you elaborate?

**Comment by foxbunny** [comment]
@ThiefMaster @tomasr8 The select list has been implemented. I tweaked the behavior a bit and made it automatically select the last option that exactly matches the keyword (case-insensitive). Other than that, it behaves pretty much as in the prototype.

**Review by ThiefMaster — COMMENTED** [review]
I tested it and it works fine AFAICT :)
Just a few small comments as usual.

File: indico/web/client/js/react/components/Select.jsx
Comment (foxbunny) [reply indico/web/client/js/react/components/Select.jsx:32]: This text isn't a placeholder though. It's an important cue that lets you tell this widget apart from a plain input. Therefore, I think it should not fail the basic contrast check for a11y. Tbh, I think placeholders shouldn't either (e.g., failing contrast check on a placeholder means important information such as date format is *imperceptible* to users with poor vision), but it's accepted as "industry standard".

File: indico/web/client/styles/partials/_sui_debt.scss
Comment (foxbunny) [reply indico/web/client/styles/partials/_sui_debt.scss:29]: Everything in that module needs to eventually be phased out and replaced with the design system. I wouldn't worry about it too much.

Comment (foxbunny) [reply indico/web/client/styles/partials/_sui_debt.scss:29]: If those values are used outside SUI, such declarations should be moved into this module. Within this module, we basically shouldn't care how dirty something is. It's not intended for long-term maintenance.

File: indico/web/client/js/custom_elements/ind_select.scss
Comment (foxbunny) [reply indico/web/client/js/custom_elements/ind_select.scss]: I think I fixed this one. It should've been `:focus-visible` instead of `:focus`.

Comment (foxbunny) [reply indico/web/client/js/custom_elements/ind_select.scss]: I'll look into why the other field receives focus, though. Cause it shouldn't. Maybe it's a Firefox bug or something.

Comment (foxbunny) [reply indico/web/client/js/custom_elements/ind_select.scss]: I don't see anything obvious in the code that would cause such behavior, so I'd say it's definitely a Firefox bug. Chrome doesn't have this issue. It doesn't affect keyboard users, and we took care of the visual glitch for mouse users, so I'm inclined to say it's good enough as it is.

File: indico/web/client/js/react/components/Select.jsx
Comment (ThiefMaster) [reply indico/web/client/js/react/components/Select.jsx:32]: TBH I find it a bit hard to distinguish between "the field has a value" and "the field has no value" right now...

File: indico/web/client/js/custom_elements/ind_select.scss
Comment (ThiefMaster) [reply indico/web/client/js/custom_elements/ind_select.scss]: works all fine now :)

File: indico/web/client/js/react/components/Select.jsx
Comment (ThiefMaster) [inline indico/web/client/js/react/components/Select.jsx:1]: I noticed this really weird behavior in Firefox where it randomly selects text when repeatedly opening and closing the dropdown... could be a Firefox bug though, so I would not spend *too* much time on it...

weird-ff-selection

Comment (ThiefMaster) [reply indico/web/client/js/react/components/Select.jsx:1]: Actually it seems to happen when moving the mouse cursor just the slightest bit while clicking.

Comment (foxbunny) [reply indico/web/client/js/react/components/Select.jsx:1]: Well, that's a common issue with any custom widget. We can mask it with CSS so at least it's not visible, but not sure if selection can be outright prevented or if it's even worth it.

File: indico/web/client/js/react/components/Select.jsx
Comment (foxbunny) [reply indico/web/client/js/react/components/Select.jsx:32]: The native select works exactly like that:

image

File: indico/web/client/js/react/components/Select.jsx
Comment (ThiefMaster) [reply indico/web/client/js/react/components/Select.jsx:1]: I never noticed it with other widgets... Would deferring the closing of the dialog help, so the click event has already been fully handled by the time it closes?

Comment (foxbunny) [reply indico/web/client/js/react/components/Select.jsx:1]: > I never noticed it with other widgets...

Happens quite often actually.

> Would deferring the closing of the dialog help, so the click event has already been fully handled by the time it closes?

Will see what I can do.

**Comment by tomasr8** [comment]
I'm getting this weird exception when I try to drag&drop the widget in the form designer (`manage/registration/XXX/form/`):

image

**Comment by foxbunny** [comment]
Ah, damn. My bad. I totally overlooked the drag & drop. It will probably affect the date picker as well. I'll do a separate PR for that.

**Comment by foxbunny** [comment]
Ok, please ignore the last fix. I'm going to try and come up with a more systematic way to address these temporary disconnects so that it can be applied to all custom elements. When I'm done, I'll first do a PR to address that in the datepicker, and then I'll use that fix in this PR as well.

**Comment by foxbunny** [comment]
This is the fix that I propose for this issue. 

https://github.com/indico/indico/pull/6447

**Comment by foxbunny** [comment]
Updated the PR to include the base class in combobox and select.

File: CHANGES.rst
Comment (ThiefMaster) [inline CHANGES.rst:123]: Please move the changelog entries to 3.3.4 (a11y section goes between bugfixes and internal changes, since there isn't one for 3.3.4 yet).

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx
Comment (ThiefMaster) [inline indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:417]: Can this still happen? I think it might be a leftover from when it was a ComboBox

File: indico/web/client/js/custom_elements/ind_select.js
Comment (ThiefMaster) [inline indico/web/client/js/custom_elements/ind_select.js:107]:
```suggestion
        setTimeout(() => {
```

File: indico/web/client/js/custom_elements/ind_select.js
Comment (ThiefMaster) [inline indico/web/client/js/custom_elements/ind_select.js:295]: `Translate.string(...)`

File: indico/web/client/js/custom_elements/ind_select.js
Comment (ThiefMaster) [inline indico/web/client/js/custom_elements/ind_select.js:228]:
```suggestion
        selectOptionWithIndex((currentIndex, optionCount) => (currentIndex + 1) % optionCount);
```

File: indico/web/client/js/custom_elements/ind_select.js
Comment (ThiefMaster) [inline indico/web/client/js/custom_elements/ind_select.js:237]:
```suggestion
        selectOptionWithIndex((currentIndex, optionCount) =>
          currentIndex <= 0 ? optionCount - 1 : currentIndex - 1
        );
```

**Review by ThiefMaster — APPROVED** [review]

**Pull request merged**
